Praise the maintainer! \o/

mueslo opened the pull request:
Some required changes:

- drop python2 compatibility

Commit:
Raise FileNotFoundError in query_scene_fname()

In case of a wrong path, this increases the length as well as the informational
value of the error message by at least nineteen.

'wat' is still raised in rarer circumstances to share the programm's confusion
with the user.

Commit:
Remove local path from mediainfo output

There is no reason to transmit this irrelevant information.  Not transmitting it
increases data parsimony, which is always a good thing.

Commit:
Use audio/video_track key 'codec_id' instead of 'codec'

Commit:
Support EAC3

Commit:
mediainfo prepends "A_" to audio codecs

Commit:
Add support for TrueHD audio

Commit:
Attemp to login 5 times before failing

Commit:
Properly format Exception messages

Commit:
Add resolution: 2160p

Commit:
Add codec: VP9

Commit:
Fix indentation

Commit:
Don't throw exception when throwing exception about unsupported codec

Commit:
Report more meaningful error if submission fails

Commit:
Show imgur URL for each uploaded image

Commit:
Retry login 10 times before giving up

Commit:
More debugging output

Commit:
Raise exception when guessit fails

Commit:
Depend on a more guessit >= 3.1

Commit:
Remove f-strings for Python <=3.5

Commit:
Limit login attempts to 4 and add delay

Commit:
Remove delay between login retries

It doesn't seem to have any effect.

Commit:
Change default cast number to 10

Prominent actors or important characters are not always listed first.

Commit:
Support video codec "avc1" and audio codec "mp4a"

Commit:
Add "confirmed_guess" that asks user

Commit:
Ask user if guessit fails to detect TV show

Commit:
Handle "Blu-Ray" source

Commit:
Add support for x265 and H.265 (Thanks, eeeeve!)

Commit:
Add support for DTS-HD and DTS:X (Thanks, eeeeve!)

Commit:
Add support for HDR10 (Thanks, eeeeve!)

Commit:
Don't prompt for mediainfo file if there's only one

Commit:
Use K/Mbps instead of K/Mb/s in mediainfo

Commit:
Add support for m2ts

Commit:
Use dict.get() for brevity

Commit:
Additional check for AVC

Commit:
Add support for VC-1 and MPEG-2 video codecs

Commit:
Overhaul _render_audio_codec()

Commit:
Add support for remux

Commit:
Fix typo

Commit:
Fix edition/proper ordering

Commit:
Move workaround for login bug to Tracker class

Commit:
Fix "'NoneType' object has no attribute 'encode'"

Looks like an actor's name can be None.

Commit:
Add todo

Commit:
Handle multiple actors separated by "/"

Not sure why, could be multiple voice actors for different languages.

Commit:
Depend on tvdb_api 2.0

Commit:
Fix TV uploads

Commit:
Prompt user for TV show selection

Commit:
Replace the installation procedure

Commit:
Fix rating for episode

Commit:
Catch ValueError if TVDB returns malformed IMDb ID

Commit:
Support MP3 as codec id "55"

Commit:
Update Installation section

Commit:
Fix pipx upgrade command

Commit:
Add pip installation option

Commit:
Fix TV show banner fallback

Commit:
Try harder to find TV show banner

Commit:
Fix adding guessit's edition information

Commit:
Fix "additional" field

Looks like guessit's "edition" can be str or sequence.

Commit:
Maybe fix a bug with autodetecting movie/tv?

mueslo commented on pythonbits/bb.py:
```diff
@@ -8,9 +8,10 @@
 
 from textwrap import dedent
-from collections import namedtuple
+from collections import namedtuple, abc
```
Note to self: incompatible with Python 2, so drop Python 2 compat

plotski replied:
What exactly do you want me to do? If that line is incompatible with Python 2, Python 2 compat is already dropped.

mueslo replied:
Ah, sorry for the misunderstanding, this is not meant for you to change, this is just a note on stuff that needs to be done after merging. So I want to merge it and then remove Python2 compatibility. But I think previous to your changes Python 2 compatibility was still there. But I don't think it's worth it to maintain since Python 2 is EOL.

mueslo commented on pythonbits/bb.py:
```diff
@@ -365,45 +424,67 @@ def _render_container(self):
-                    'x264' in video_track['writing_library']):
+        if (video_track['codec_id'] in ('V_MPEG4/ISO/AVC', 'AVC', 'avc1') or
+            video_track['format'] == 'AVC'):
+            if 'x264' in video_track.get('writing_library', ''):
                 return 'x264'
             else:
                 return 'H.264'
-        elif video_track['codec'] == 'XVID':
+        elif video_track['codec_id'] in ('V_MPEGH/ISO/HEVC', 'HEVC'):
+            if 'x265' in video_track.get('writing_library', ''):
+                return 'x265'
+            else:
+                return 'H.265'
```
Make style consistent

plotski replied:
Looks perfectly readable to me. I don't see any inconsistencies.

mueslo replied:
You're right, I'm stupid. I misread the diff. It's good.

mueslo commented on pythonbits/bb.py:
```diff
@@ -209,6 +218,31 @@ class VideoSubmission(BbSubmission):
     def _render_guess(self):
         return {k: v for k, v in guessit.guessit(self['path']).items()}
 
+    def _render_confirmed_guess(self):
+        guess = {k: v for k, v in guessit.guessit(self['path']).items()}
+        print('Confirm by pressing enter or type in correct values:')
+
+        guess['title'] = input('Guessed title: %s > ' % (guess['title'],)) or guess['title']
+
+        confirmed_type = None
+        while confirmed_type not in ('movie', 'episode'):
+            user_type = 'tv' if guess['type'] == 'episode' else guess['type']
+            confirmed_type = (input('Guessed type ("movie" or "tv"): %s > ' % (user_type,))
+                              or guess['type'])
+            confirmed_type = 'episode' if confirmed_type == 'tv' else confirmed_type
+        guess['type'] = confirmed_type
+
+        if guess['type'] == 'episode' and 'season' not in guess:
+            while True:
+                try:
+                    guess['season'] = int(input('Season > '))
+                except (ValueError, TypeError):
+                    pass
+                else:
+                    break
+
+        return guess
+
```
mhhhh

plotski replied:
It's possible the "confirmed_guess" stuff can be removed. I'll look into it over the weekend.

mueslo replied:
I'm not too sure if the use case for this is really there, I will check it out tonight.

plotski replied:
I think I've added this as a quick and dirty fix when guessit was confused/bugged, but I don't remember much else. I'd like to have something like this because I think it's important to not rely on autodetection, but this implementation is probably bad. Ideally we would have a TUI form that is prefilled and can be edited by the user if necessary.

plotski replied:
OK, I figured it out. TV show packs that don't specify a season or do it badly are detected as "movie" and things break horribly. I've made a few commits that clean this up. The "confirmed_guess" thing is still pretty dirty, but it works and allows easy recovery from otherwise fatal misdetection. Please let's keep it until someone has time to improve.

mueslo replied:
Sorry, but if that is the point, then the fix should be in `VideoSubmission.subcategory()`. This adds a lot of needless complexity. But also if it were implemented there, how often does this realistically happen? And how much time is lost by having to confirm this every time vs. just rerunning it with a manual title? I have removed it in the merged commits. 

If a TV show pack doesn't specify a season you can just run it as e.g. `pythonbits /some/strange/path "breaking bad s01"`

plotski replied:
`confirmed_request` is only used if the info doesn't make sense, e.g. tv show but no season. Otherwise it has no effect. This happened at least twice to me, and once is enough to make this a bug in my opinion.

But I didn't realize I can put info in the TITLE argument. I thought it was literally the show/movie title. I liked the idea of requesting additional info from the user ad hoc, but I guess this is also a fix.

mueslo commented on pythonbits/bb.py:
```diff
@@ -73,7 +79,10 @@ def subcategorise(self):
     def submit(payload):
         t = Tracker()
-        return t.upload(**payload)
+        try:
+            return t.upload(**payload)
+        except TrackerException as e:
+            log.error('Upload failed: %s' % (e,))
```
Why prevent the error from being raised?

plotski replied:
Because throwing tracebacks at the user isn't very nice. But I guess errors should be caught somewhere in __main__(). I'll look into it.

plotski replied:
I removed that exception catcher here. `__main__` is actually catching exceptions, it's just printing it with the traceback. Maybe I'll try to tell logbook to only print the traceback when debugging is enabled or something.

mueslo commented on pythonbits/bb.py:
```diff
@@ -239,13 +274,25 @@ def _render_tv_specifier(self):
             return TvSpecifier(title, guess['season'],
                                guess.get('episode', None))
 
+        cguess = self['confirmed_guess']
+        log.debug('Confirmed guess: %s' % (cguess,))
+        if cguess['type'] == 'episode':
+            if self['title_arg']:
+                title = self['title_arg']
+            else:
+                title = cguess['title']
+            return TvSpecifier(title, cguess['season'],
+                               cguess.get('episode', None))
+
```
not sure about the point of this

plotski replied:
I don't really remember, but I think there was a problem with auto-detecting stuff. (Wish I had the time to write tests for everything.)

mueslo replied:
> 
> Wish I had the time to write tests for everything.

Ditto... Maybe I can get the tests working again.

mueslo commented on pythonbits/bb.py:
```diff
@@ -365,45 +424,67 @@ def _render_container(self):
-        if video_track['codec'] in ('V_MPEG4/ISO/AVC', 'AVC'):
-            if ('writing_library' in video_track and
-                    'x264' in video_track['writing_library']):
+        if (video_track['codec_id'] in ('V_MPEG4/ISO/AVC', 'AVC', 'avc1') or
```
Still need to find the reason for this one.

Commit:
Don't catch TrackerException in BBSubmission.submit()

plotski committed:
Merge branch 'master' into master

Commit:
Catch LookError from imdbpie.get_title_ratings()

This should fix crashes with the following error:

    LookupError: Title not found. ttxxxxxxx is a redirection imdb id

Commit:
Use https for IMDb link

Commit:
Map region codes to movie titles in ImdbResult.summary["titles"]

Commit:
Set "... AKA ..." title (fixes #47)

plotski commented on README.md:
```diff
@@ -1,24 +1,19 @@
-  - you can also manually specify things such as the torrent file or screenshots, this will prevent the programs from being called, removing the dependency
-
-If you don't want to use a virtualenv but keep system pollution with PyPI packages to a minimum, install via `pip install --user`. For more information, visit [this site](https://packaging.python.org/guides/installing-using-pip-and-virtualenv/).
+This is a sloppy fork of [pythonBits](https://github.com/mueslo/pythonbits) that
```
You probably don't want to pull my README edits.

mueslo replied:
Yep. I will pull manually on my local branch, and cherry-pick-delete some commits.

Commit:
Better way to detect MP3 (hopefully)

Commit:
Default to previous method of detecting MP3

Commit:
Pass --category to guessit

Commit:
Add some comments to _render_confirmed_guess()

Commit:
Deduplicate code and prevent KeyError('season')

plotski commented:
I think I've fixed everything. Let me know if I missed something.

Any chance you might be able to pull and make a new release any time soon?

Commit:
Get international titles for TV shows from IMDb

TVDB doesn't seem to know provide titles.

Commit:
Add IMDB.get_info()

This allows us to get an IMDBResult if we already know the IMDb ID.

Commit:
Add support for "... AKA ..." in title for TV shows

mueslo commented:
> I think I've fixed everything. Let me know if I missed something.
> 
> Any chance you might be able to pull and make a new release any time soon?

Yep, I've pulled locally and am merging and testing right now.
